File: libs/moqtail-ts/src/model/control/subscribe.ts
import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from '../common/byte_buffer'
import { Location } from '../common/location'
import { Tuple } from '../common/tuple'
import { KeyValuePair } from '../common/pair'
import { ControlMessageType, FilterType, GroupOrder } from '../control/constant'
import { FullTrackName } from '../data'

export class Subscribe {
  private constructor(
    public requestId: bigint,
    public trackAlias: bigint,
    public fullTrackName: FullTrackName,
    public subscriberPriority: number,
    public groupOrder: GroupOrder,
    public forward: boolean,
    public filterType: FilterType,
    public startLocation: Location | undefined,
    public endGroup: bigint | undefined,
    public subscribeParameters: KeyValuePair[],
  ) {}

  static newNextGroupStart(
    requestId: bigint,
    trackAlias: bigint,
    fullTrackName: FullTrackName,
    subscriberPriority: number,
    groupOrder: GroupOrder,
    forward: boolean,
    subscribeParameters: KeyValuePair[],
  ): Subscribe {
    return new Subscribe(
      requestId,
      trackAlias,
      fullTrackName,
      subscriberPriority,
      groupOrder,
      forward,
      FilterType.NextGroupStart,
      undefined,
      undefined,
      subscribeParameters,
    )
  }

  static newLatestObject(
    requestId: bigint,
    trackAlias: bigint,
    fullTrackName: FullTrackName,
    subscriberPriority: number,
    groupOrder: GroupOrder,
    forward: boolean,
    subscribeParameters: KeyValuePair[],
  ): Subscribe {
    return new Subscribe(
      requestId,
      trackAlias,
      fullTrackName,
      subscriberPriority,
      groupOrder,
      forward,
      FilterType.LatestObject,
      undefined,
      undefined,
      subscribeParameters,
    )
  }

  static newAbsoluteStart(
    requestId: bigint,
    trackAlias: bigint,
    fullTrackName: FullTrackName,
    subscriberPriority: number,
    groupOrder: GroupOrder,
    forward: boolean,
    startLocation: Location,
    subscribeParameters: KeyValuePair[],
  ): Subscribe {
    return new Subscribe(
      requestId,
      trackAlias,
      fullTrackName,
      subscriberPriority,
      groupOrder,
      forward,
      FilterType.AbsoluteStart,
      startLocation,
      undefined,
      subscribeParameters,
    )
  }

  static newAbsoluteRange(
    requestId: bigint,
    trackAlias: bigint,
    fullTrackName: FullTrackName,
    subscriberPriority: number,
    groupOrder: GroupOrder,
    forward: boolean,
    startLocation: Location,
    endGroup: bigint,
    subscribeParameters: KeyValuePair[],
  ): Subscribe {
    if (endGroup < startLocation.group) {
      throw new Error('End Group must be >= Start Group')
    }
    return new Subscribe(
      requestId,
      trackAlias,
      fullTrackName,
      subscriberPriority,
      groupOrder,
      forward,
      FilterType.AbsoluteRange,
      startLocation,
      endGroup,
      subscribeParameters,
    )
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(ControlMessageType.Subscribe)

    const payload = new ByteBuffer()
    payload.putVI(this.requestId)
    payload.putVI(this.trackAlias)
    payload.putBytes(this.fullTrackName.serialize().toUint8Array())
    payload.putU8(this.subscriberPriority)
    payload.putU8(this.groupOrder)
    payload.putU8(this.forward ? 1 : 0)
    payload.putVI(this.filterType)

    if (this.filterType === FilterType.AbsoluteStart || this.filterType === FilterType.AbsoluteRange) {
      if (!this.startLocation) {
        throw new Error('StartLocation required for selected filterType')
      }
      payload.putLocation(this.startLocation)
    }

    if (this.filterType === FilterType.AbsoluteRange) {
      if (this.endGroup == null) {
        throw new Error('EndGroup required for AbsoluteRange')
      }
      payload.putVI(this.endGroup)
    }

    payload.putVI(this.subscribeParameters.length)
    for (const param of this.subscribeParameters) {
      payload.putBytes(param.serialize().toUint8Array())
    }

    const payloadBytes = payload.toUint8Array()
    buf.putU16(payloadBytes.length)
    buf.putBytes(payloadBytes)

    return buf.freeze()
  }

  static parsePayload(buf: BaseByteBuffer): Subscribe {
    const requestId = buf.getVI()
    const trackAlias = buf.getVI()
    const fullTrackName = buf.getFullTrackName()
    const subscriberPriority = buf.getU8()
    const groupOrder = buf.getU8()
    const forward = buf.getU8() === 1
    const filterType = Number(buf.getVI()) as FilterType

    let startLocation: Location | undefined = undefined
    let endGroup: bigint | undefined = undefined

    if (filterType === FilterType.AbsoluteStart || filterType === FilterType.AbsoluteRange) {
      startLocation = buf.getLocation()
    }
    if (filterType === FilterType.AbsoluteRange) {
      endGroup = buf.getVI()
    }

    const paramCount = Number(buf.getVI())
    const subscribeParameters: KeyValuePair[] = []
    for (let i = 0; i < paramCount; i++) {
      subscribeParameters.push(KeyValuePair.deserialize(buf))
    }

    return new Subscribe(
      requestId,
      trackAlias,
      fullTrackName,
      subscriberPriority,
      groupOrder,
      forward,
      filterType,
      startLocation,
      endGroup,
      subscribeParameters,
    )
  }
}

if (import.meta.vitest) {
  const { describe, it, expect } = import.meta.vitest

  function buildTestSubscribe(): Subscribe {
    return Subscribe.newAbsoluteRange(
      128242n,
      999n,
      FullTrackName.tryNew('track/namespace', 'trackName'),
      31,
      GroupOrder.Original,
      true,
      new Location(81n, 81n),
      100n,
      [KeyValuePair.tryNewVarInt(0n, 10n), KeyValuePair.tryNewBytes(1n, new TextEncoder().encode('DemoString'))],
    )
  }

  describe('Subscribe', () => {
    it('should roundtrip correctly', () => {
      const subscribe = buildTestSubscribe()
      const serialized = subscribe.serialize()

      const buf = new ByteBuffer()
      buf.putBytes(serialized.toUint8Array())
      const msgType = buf.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.Subscribe))

      const msgLength = buf.getU16()
      expect(msgLength).toBe(buf.remaining)

      const deserialized = Subscribe.parsePayload(buf)
      expect(deserialized).toEqual(subscribe)
      expect(buf.remaining).toBe(0)
    })

    it('should roundtrip with excess trailing bytes', () => {
      const subscribe = buildTestSubscribe()
      const serialized = subscribe.serialize()
      const extra = new Uint8Array([...serialized.toUint8Array(), 9, 1, 1])

      const buf = new ByteBuffer()
      buf.putBytes(extra)

      const msgType = buf.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.Subscribe))

      const msgLength = buf.getU16()
      expect(msgLength).toBe(buf.remaining - 3)

      const deserialized = Subscribe.parsePayload(buf)
      expect(deserialized).toEqual(subscribe)

      const trailing = buf.toUint8Array().slice(buf.offset)
      expect(Array.from(trailing)).toEqual([9, 1, 1])
    })

    describe('Subscribe Constructors', () => {
      it('should create a Subscribe with AbsoluteRange filter', () => {
        const subscribe = Subscribe.newAbsoluteRange(
          128242n,
          999n,
          FullTrackName.tryNew('track/namespace', 'trackName'),
          31,
          GroupOrder.Original,
          true,
          new Location(81n, 81n),
          100n,
          [],
        )

        expect(subscribe.filterType).toBe(FilterType.AbsoluteRange)
        expect(subscribe.startLocation).toEqual(new Location(81n, 81n))
        expect(subscribe.endGroup).toBe(100n)
      })

      it('should throw an error if EndGroup < StartGroup', () => {
        expect(() =>
          Subscribe.newAbsoluteRange(
            128242n,
            999n,
            FullTrackName.tryNew('track/namespace', 'trackName'),
            31,
            GroupOrder.Original,
            true,
            new Location(81n, 81n),
            80n,
            [],
          ),
        ).toThrow('End Group must be >= Start Group')
      })
    })

    it('should throw on invalid filterType', () => {
      const buf = new ByteBuffer()
      buf.putVI(ControlMessageType.Subscribe)
      buf.putVI(128242n)
      buf.putVI(999n)
      buf.putTuple(Tuple.fromUtf8Path('invalid/filter'))
      buf.putVI(10)
      buf.putBytes(new TextEncoder().encode('InvalidTest'))
      buf.putU8(31)
      buf.putU8(GroupOrder.Original)
      buf.putU8(1)
      buf.putVI(9999)

      expect(() => Subscribe.parsePayload(buf)).toThrow()
    })
    it('should handle empty subscribeParameters', () => {
      const subscribe = Subscribe.newLatestObject(
        128242n,
        999n,
        FullTrackName.tryNew('track/namespace', 'trackName'),
        31,
        GroupOrder.Original,
        true,
        [],
      )

      const serialized = subscribe.serialize()
      const buf = new ByteBuffer()
      buf.putBytes(serialized.toUint8Array())

      const msgType = buf.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.Subscribe))
      buf.getU16()

      const deserialized = Subscribe.parsePayload(buf)
      expect(deserialized).toEqual(subscribe)
      expect(buf.remaining).toBe(0)
    })

    it('should throw on partial message', () => {
      const subscribe = buildTestSubscribe()
      const serialized = subscribe.serialize()
      const serializedBytes = serialized.toUint8Array()

      const partialBytes = serializedBytes.slice(0, Math.floor(serializedBytes.length / 2))
      const buf = new ByteBuffer()
      buf.putBytes(partialBytes)

      try {
        buf.getVI()
        buf.getU16()

        expect(() => Subscribe.parsePayload(buf)).toThrow()
      } catch (err) {
        expect(err).toBeInstanceOf(Error)
      }
    })
  })
}

File: libs/moqtail-rs/src/model/control/subscribe.rs
use super::constant::{ControlMessageType, FilterType, GroupOrder};
use super::control_message::ControlMessageTrait;
use crate::model::common::location::Location;
use crate::model::common::pair::KeyValuePair;
use crate::model::common::tuple::Tuple;
use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::error::ParseError;
use bytes::{Buf, BufMut, Bytes, BytesMut};

#[derive(Debug, PartialEq, Clone)]
pub struct Subscribe {
  pub request_id: u64,
  pub track_alias: u64,
  pub track_namespace: Tuple,
  pub track_name: String,
  pub subscriber_priority: u8,
  pub group_order: GroupOrder,
  pub forward: bool,
  pub filter_type: FilterType,
  pub start_location: Option<Location>,
  pub end_group: Option<u64>,
  // TODO: make the following optional
  pub subscribe_parameters: Vec<KeyValuePair>,
}

#[allow(clippy::too_many_arguments)]
impl Subscribe {
  pub fn new_next_group_start(
    request_id: u64,
    track_alias: u64,
    track_namespace: Tuple,
    track_name: String,
    subscriber_priority: u8,
    group_order: GroupOrder,
    forward: bool,
    subscribe_parameters: Vec<KeyValuePair>,
  ) -> Self {
    Self {
      request_id,
      track_alias,
      track_namespace,
      track_name,
      subscriber_priority,
      group_order,
      forward,
      filter_type: FilterType::NextGroupStart,
      start_location: None,
      end_group: None,
      subscribe_parameters,
    }
  }

  pub fn new_latest_object(
    request_id: u64,
    track_alias: u64,
    track_namespace: Tuple,
    track_name: String,
    subscriber_priority: u8,
    group_order: GroupOrder,
    forward: bool,
    subscribe_parameters: Vec<KeyValuePair>,
  ) -> Self {
    Self {
      request_id,
      track_alias,
      track_namespace,
      track_name,
      subscriber_priority,
      group_order,
      forward,
      filter_type: FilterType::LatestObject,
      start_location: None,
      end_group: None,
      subscribe_parameters,
    }
  }

  pub fn new_absolute_start(
    request_id: u64,
    track_alias: u64,
    track_namespace: Tuple,
    track_name: String,
    subscriber_priority: u8,
    group_order: GroupOrder,
    forward: bool,
    start_location: Location,
    subscribe_parameters: Vec<KeyValuePair>,
  ) -> Self {
    Self {
      request_id,
      track_alias,
      track_namespace,
      track_name,
      subscriber_priority,
      group_order,
      forward,
      filter_type: FilterType::AbsoluteStart,
      start_location: Some(start_location),
      end_group: None,
      subscribe_parameters,
    }
  }

  pub fn new_absolute_range(
    request_id: u64,
    track_alias: u64,
    track_namespace: Tuple,
    track_name: String,
    subscriber_priority: u8,
    group_order: GroupOrder,
    forward: bool,
    start_location: Location,
    end_group: u64,
    subscribe_parameters: Vec<KeyValuePair>,
  ) -> Self {
    assert!(
      end_group >= start_location.group,
      "End Group must be >= Start Group"
    );
    Self {
      request_id,
      track_alias,
      track_namespace,
      track_name,
      subscriber_priority,
      group_order,
      forward,
      filter_type: FilterType::AbsoluteRange,
      start_location: Some(start_location),
      end_group: Some(end_group),
      subscribe_parameters,
    }
  }
}
impl ControlMessageTrait for Subscribe {
  fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(ControlMessageType::Subscribe)?;

    let mut payload = BytesMut::new();
    payload.put_vi(self.request_id)?;
    payload.put_vi(self.track_alias)?;

    payload.extend_from_slice(&self.track_namespace.serialize()?);
    payload.put_vi(self.track_name.len())?;
    payload.extend_from_slice(self.track_name.as_bytes());
    payload.put_u8(self.subscriber_priority);
    payload.put_u8(self.group_order as u8);
    payload.put_u8(if self.forward { 1 } else { 0 });
    payload.put_vi(self.filter_type)?;

    match self.filter_type {
      FilterType::AbsoluteStart => {
        if let Some(ref loc) = self.start_location {
          payload.extend_from_slice(&loc.serialize()?);
        } else {
          unreachable!()
        }
      }
      FilterType::AbsoluteRange => {
        if let Some(ref loc) = self.start_location {
          payload.extend_from_slice(&loc.serialize()?);
        }
        if let Some(eg) = self.end_group {
          payload.put_vi(eg)?;
        }
      }
      _ => {}
    }

    payload.put_vi(self.subscribe_parameters.len())?;
    for param in &self.subscribe_parameters {
      payload.extend_from_slice(&param.serialize()?);
    }

    let payload_len: u16 = payload
      .len()
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "Subscribe::serialize",
        from_type: "usize",
        to_type: "u16",
        details: e.to_string(),
      })?;

    buf.put_u16(payload_len);
    buf.extend_from_slice(&payload);
    Ok(buf.freeze())
  }

  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError> {
    let request_id = payload.get_vi()?;
    let track_alias = payload.get_vi()?;
    let track_namespace = Tuple::deserialize(payload)?;

    let name_len_u64 = payload.get_vi()?;
    let name_len: usize = name_len_u64
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "Subscribe::parse_payload(track_name_len)",
        from_type: "u64",
        to_type: "usize",
        details: e.to_string(),
      })?;

    if payload.remaining() < name_len {
      return Err(ParseError::NotEnoughBytes {
        context: "Subscribe::parse_payload(track_name)",
        needed: name_len,
        available: payload.remaining(),
      });
    }
    let track_name_bytes = payload.copy_to_bytes(name_len);
    let track_name =
      String::from_utf8(track_name_bytes.to_vec()).map_err(|e| ParseError::InvalidUTF8 {
        context: "Subscribe::parse_payload(track_name)",
        details: e.to_string(),
      })?;

    if payload.remaining() < 1 {
      return Err(ParseError::NotEnoughBytes {
        context: "Subscribe::parse_payload(subscriber_priority)",
        needed: 1,
        available: 0,
      });
    }
    let subscriber_priority = payload.get_u8();

    if payload.remaining() < 1 {
      return Err(ParseError::NotEnoughBytes {
        context: "Subscribe::parse_payload(group_order)",
        needed: 1,
        available: 0,
      });
    }
    let group_order_raw = payload.get_u8();
    let group_order = GroupOrder::try_from(group_order_raw)?;

    let forward_raw = payload.get_u8();
    let forward = match forward_raw {
      0 => false,
      1 => true,
      _ => {
        return Err(ParseError::ProcotolViolation {
          context: "Subscribe::parse_payload(forward)",
          details: format!("Invalid value: {forward_raw}"),
        });
      }
    };

    let filter_type_raw = payload.get_vi()?;
    let filter_type = FilterType::try_from(filter_type_raw)?;

    let mut start_location: Option<Location> = None;
    let mut end_group: Option<u64> = None;

    match filter_type {
      FilterType::AbsoluteRange => {
        start_location = Some(Location::deserialize(payload)?);
        end_group = Some(payload.get_vi()?);
      }
      FilterType::AbsoluteStart => {
        start_location = Some(Location::deserialize(payload)?);
      }
      FilterType::LatestObject => {}
      FilterType::NextGroupStart => {}
    }

    let param_count_u64 = payload.get_vi()?;
    let param_count: usize =
      param_count_u64
        .try_into()
        .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
          context: "Subscribe::deserialize(param_count)",
          from_type: "u64",
          to_type: "usize",
          details: e.to_string(),
        })?;

    let mut subscribe_parameters = Vec::with_capacity(param_count);
    for _ in 0..param_count {
      let param = KeyValuePair::deserialize(payload)?;
      subscribe_parameters.push(param);
    }

    Ok(Box::new(Subscribe {
      request_id,
      track_alias,
      track_namespace,
      track_name,
      subscriber_priority,
      group_order,
      forward,
      filter_type,
      start_location,
      end_group,
      subscribe_parameters,
    }))
  }
  fn get_type(&self) -> ControlMessageType {
    ControlMessageType::Subscribe
  }
}
#[cfg(test)]
mod tests {
  use super::*;
  use bytes::Buf;

  #[test]
  fn test_roundtrip() {
    let request_id = 128242;
    let track_alias = 999;
    let track_namespace = Tuple::from_utf8_path("nein/nein/nein");
    let track_name = "${Name}".to_string();
    let subscriber_priority = 31;
    let group_order = GroupOrder::Original;
    let forward = true;
    let filter_type = FilterType::AbsoluteRange;
    let start_location = Location {
      group: 81,
      object: 81,
    };
    let end_group = 25;
    let subscribe_parameters = vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"I'll sync you up")).unwrap(),
    ];
    let subscribe = Subscribe {
      request_id,
      track_alias,
      track_namespace,
      track_name,
      subscriber_priority,
      group_order,
      forward,
      filter_type,
      start_location: Some(start_location),
      end_group: Some(end_group),
      subscribe_parameters,
    };

    let mut buf = subscribe.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::Subscribe as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = Subscribe::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, subscribe);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_excess_roundtrip() {
    let request_id = 128242;
    let track_alias = 999;
    let track_namespace = Tuple::from_utf8_path("nein/nein/nein");
    let track_name = "${Name}".to_string();
    let subscriber_priority = 31;
    let group_order = GroupOrder::Original;
    let forward = true;
    let filter_type = FilterType::AbsoluteRange;
    let start_location = Location {
      group: 81,
      object: 81,
    };
    let end_group = 25;
    let subscribe_parameters = vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"I'll sync you up")).unwrap(),
    ];
    let subscribe = Subscribe {
      request_id,
      track_alias,
      track_namespace,
      track_name,
      subscriber_priority,
      group_order,
      forward,
      filter_type,
      start_location: Some(start_location),
      end_group: Some(end_group),
      subscribe_parameters,
    };

    let serialized = subscribe.serialize().unwrap();
    let mut excess = BytesMut::new();
    excess.extend_from_slice(&serialized);
    excess.extend_from_slice(&[9u8, 1u8, 1u8]);
    let mut buf = excess.freeze();

    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::Subscribe as u64);
    let msg_length = buf.get_u16();

    assert_eq!(msg_length as usize, buf.remaining() - 3);
    let deserialized = Subscribe::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, subscribe);
    assert_eq!(buf.chunk(), &[9u8, 1u8, 1u8]);
  }

  #[test]
  fn test_partial_message() {
    let request_id = 128242;
    let track_alias = 999;
    let track_namespace = Tuple::from_utf8_path("nein/nein/nein");
    let track_name = "${Name}".to_string();
    let subscriber_priority = 31;
    let group_order = GroupOrder::Original;
    let forward = true;
    let filter_type = FilterType::AbsoluteRange;
    let start_location = Location {
      group: 81,
      object: 81,
    };
    let end_group = 25;
    let subscribe_parameters = vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"I'll sync you up")).unwrap(),
    ];
    let subscribe = Subscribe {
      request_id,
      track_alias,
      track_namespace,
      track_name,
      subscriber_priority,
      group_order,
      forward,
      filter_type,
      start_location: Some(start_location),
      end_group: Some(end_group),
      subscribe_parameters,
    };

    let mut buf = subscribe.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::Subscribe as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());

    let upper = buf.remaining() / 2;
    let mut partial = buf.slice(..upper);
    let deserialized = Subscribe::parse_payload(&mut partial);
    assert!(deserialized.is_err());
  }
}
